backend/pkg/modules/internal/differ/spec_diffs.go

```diff
@@ -0,0 +1,277 @@
+
+	apiEvent := event.APIEvent
+	specKey := _speculator.GetSpecKey(apiEvent.HostSpecName, strconv.Itoa(int(apiEvent.DestinationPort)))
+	speculatorAccessor := p.accessor.GetSpeculatorAccessor()
+
+	if !speculatorAccessor.HasProvidedSpec(specKey) && !speculatorAccessor.HasApprovedSpec(specKey) {
+		log.Debugf("No diffs to calculate")
+		return
+	}
+
+	speculatorTelemetry := speculatorutils.ConvertModelsToSpeculatorTelemetry(event.Telemetry)
+
+	reconstructedDiffType := models.DiffTypeNODIFF
+	providedDiffType := models.DiffTypeNODIFF
+	if speculatorAccessor.HasProvidedSpec(specKey) {
+		// calculate diffs base on the event
+		providedDiff, err = speculatorAccessor.DiffTelemetry(speculatorTelemetry, _spec.DiffSourceProvided)
+		if err != nil {
+			log.Errorf("Failed to diff telemetry against provided spec: %v", err)
+			return
+		}
+		if err := setAPIEventProvidedDiff(apiEvent, providedDiff); err != nil {
+			log.Errorf("Failed to set api event provided diff: %v", err)
+			return
+		}
+		providedDiffType = convertAPIDiffType(providedDiff.Type)
+	} else if speculatorAccessor.HasApprovedSpec(specKey) {
+		// calculate diffs base on the event
+		reconstructedDiff, err = speculatorAccessor.DiffTelemetry(speculatorTelemetry, _spec.DiffSourceReconstructed)
+		if err != nil {
+			log.Errorf("Failed to diff telemetry against approved spec: %v", err)
+			return
+		}
+		if err := setAPIEventReconstructedDiff(apiEvent, reconstructedDiff); err != nil {
+			log.Errorf("Failed to set api event reconstructed diff: %v", err)
+			return
+		}
+		reconstructedDiffType = convertAPIDiffType(reconstructedDiff.Type)
+	}
+
+	apiEvent.SpecDiffType = getHighestPrioritySpecDiffType(providedDiffType, reconstructedDiffType)
+
+	// save api event with diffs in db
+	if err := p.accessor.UpdateAPIEvent(ctx, apiEvent); err != nil {
+		log.Errorf("Failed to update api event: %v", err)
+		return
+	}
+
+	p.setDiffEventToSend(apiEvent)
+}
+
+func (p *pluginDiffer) setDiffEventToSend(event *database.APIEvent) {
+	if event.SpecDiffType == models.DiffTypeNODIFF {
+		return
+	}
+	if p.getTotalEvents() > 500 {
+		log.Warn("Diff events threshold reached (500), ignoring event")
+		return
+	}
+
+	var newSpec string
+	var oldSpec string
+	var hash [32]byte
+	if event.HasReconstructedSpecDiff {
+		newSpec = event.NewReconstructedSpec
+		oldSpec = event.OldReconstructedSpec
+	} else if event.HasProvidedSpecDiff {
+		newSpec = event.NewProvidedSpec
+		oldSpec = event.OldProvidedSpec
+	}
+
+	hash = sha256.Sum256([]byte(newSpec + oldSpec))
+	t := time.Time(event.Time)
+
+	p.Lock()
+	defer p.Unlock()
+	if len(p.apiIDToDiffs[event.APIInfoID]) == 0 {
+		p.apiIDToDiffs[event.APIInfoID] = make(map[[32]byte]global.Diff)
+	}
+	if _, ok := p.apiIDToDiffs[event.APIInfoID][hash]; !ok {
+		p.increaseTotalEvents()
+	}
+	p.apiIDToDiffs[event.APIInfoID][hash] = global.Diff{
+		DiffType: convertDiffType(event.SpecDiffType),
+		LastSeen: t.Unix(),
+		NewSpec:  newSpec,
+		OldSpec:  oldSpec,
+	}
```

Comment:
Diffs reported at the API level should also be set as API level annotation, this way they can be displayed in APIClarity UI in the API inventory section in a dedicated diff tab as we do in Panoptica.